ukvm/ukvm_gdb_kvm_x86_64.c

```diff
@@ -0,0 +1,460 @@
+
+    gregs->rax = kregs.rax;
+    gregs->rbx = kregs.rbx;
+    gregs->rcx = kregs.rcx;
+    gregs->rdx = kregs.rdx;
+
+    gregs->rsi = kregs.rsi;
+    gregs->rdi = kregs.rdi;
+    gregs->rbp = kregs.rbp;
+    gregs->rsp = kregs.rsp;
+
+    gregs->r8 = kregs.r8;
+    gregs->r9 = kregs.r9;
+    gregs->r10 = kregs.r10;
+    gregs->r11 = kregs.r11;
+
+    gregs->rip = kregs.rip;
+    gregs->eflags = kregs.rflags;
+
+    gregs->cs = sregs.cs.selector;
+    gregs->ss = sregs.ss.selector;
+    gregs->ds = sregs.ds.selector;
+    gregs->es = sregs.es.selector;
+    gregs->fs = sregs.fs.selector;
+    gregs->gs = sregs.gs.selector;
+
+    return 0;
+}
+
+/*
+ * Writes the shadow registers from a stream of hexadecimal digits for each register in
+ * GDB register order, where each register is in target endian order.
+ */
+int ukvm_gdb_write_registers(struct ukvm_hv *hv,
+                             uint8_t *registers,
+                             size_t len)
+{
+    struct kvm_regs kregs;
+    struct kvm_sregs sregs;
+    struct ukvm_gdb_regs *gregs = (struct ukvm_gdb_regs *) registers;
+    int ret;
+
+    /* Let's read all registers just in case we miss filling one of them. */
+    ret = ioctl(hv->b->vcpufd, KVM_GET_REGS, &kregs);
+    if (ret == -1) {
+        err(1, "KVM_GET_REGS");
+        return -1;
+    }
+
+    ret = ioctl(hv->b->vcpufd, KVM_GET_SREGS, &sregs);
+    if (ret == -1) {
+        err(1, "KVM_GET_REGS");
+        return -1;
+    }
+
+    if (len < sizeof(struct ukvm_gdb_regs))
+        return -1;
+
+    kregs.rax = gregs->rax;
+    kregs.rbx = gregs->rbx;
+    kregs.rcx = gregs->rcx;
+    kregs.rdx = gregs->rdx;
+
+    kregs.rsi = gregs->rsi;
+    kregs.rdi = gregs->rdi;
+    kregs.rbp = gregs->rbp;
+    kregs.rsp = gregs->rsp;
+
+    kregs.r8 = gregs->r8;
+    kregs.r9 = gregs->r9;
+    kregs.r10 = gregs->r10;
+    kregs.r11 = gregs->r11;
+
+    kregs.rip = gregs->rip;
+    kregs.rflags = gregs->eflags;
+
+    /* XXX: not sure if just setting .selector is enough. */
+    sregs.cs.selector = gregs->cs;
+    sregs.ss.selector = gregs->ss;
+    sregs.ds.selector = gregs->ds;
+    sregs.es.selector = gregs->es;
+    sregs.fs.selector = gregs->fs;
+    sregs.gs.selector = gregs->gs;
+
+    ret = ioctl(hv->b->vcpufd, KVM_SET_REGS, &kregs);
+    if (ret == -1) {
+        err(1, "KVM_GET_REGS");
+        return -1;
+    }
+
+    ret = ioctl(hv->b->vcpufd, KVM_SET_SREGS, &sregs);
+    if (ret == -1) {
+        err(1, "KVM_GET_REGS");
+        return -1;
+    }
+
+    return 0;
+}
+
+int ukvm_gdb_add_breakpoint(struct ukvm_hv *hv, uint32_t type,
+                            ukvm_gpa_t addr, size_t len)
+{
+    struct breakpoint_t *bp;
+
+    if (bp_list_find(type, addr, len))
+        return 0;
+
+    bp = bp_list_insert(type, addr, len);
+    if (bp == NULL)
+        return -1;
+
+    if (type == GDB_BREAKPOINT_SW)
+        kvm_arch_insert_sw_breakpoint(hv, bp);
+
+    if (ukvm_gdb_update_guest_debug(hv) < 0)
```

Comment:
I'd prefer to use `== -1` rather than `< 0` in checks like these and more below.